File: app/src/main/java/com/arnaudpiroelle/manga/provider/japscan/downloader/JapScanDownloader.java
package com.arnaudpiroelle.manga.provider.japscan.downloader;


import android.support.annotation.VisibleForTesting;

import com.arnaudpiroelle.manga.core.provider.MangaProvider;
import com.arnaudpiroelle.manga.core.utils.HttpUtils;
import com.arnaudpiroelle.manga.model.Chapter;
import com.arnaudpiroelle.manga.model.Manga;
import com.arnaudpiroelle.manga.model.Page;
import com.arnaudpiroelle.manga.provider.japscan.api.JapScanApiService;
import com.arnaudpiroelle.manga.provider.japscan.api.JapScanDataApiService;

import java.io.InputStream;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import retrofit.client.Response;

import static com.arnaudpiroelle.manga.model.Chapter.ChapterBuilder.createChapter;
import static com.arnaudpiroelle.manga.model.Manga.MangaBuilder.createManga;
import static com.arnaudpiroelle.manga.model.Page.PageBuilder.createPage;

public class JapScanDownloader implements MangaProvider {

    private JapScanApiService japScanApiService;
    private JapScanDataApiService japScanDataApiService;

    public JapScanDownloader(JapScanApiService japScanApiService, JapScanDataApiService japScanDataApiService) {
        this.japScanApiService = japScanApiService;
        this.japScanDataApiService = japScanDataApiService;
    }

    public List<Manga> findMangas() {
        Response mangaListResponse = japScanApiService.getMangaList();

        String body = HttpUtils.convertFrom(mangaListResponse);

        return parseMangaList(body);
    }

    public List<Chapter> findChapters(Manga manga) {
        Response mangaResponse = japScanApiService.getManga(manga.getMangaAlias());

        String body = HttpUtils.convertFrom(mangaResponse);

        List<Chapter> list = parseMangaChapters(body);
        Collections.reverse(list);

        return list;
    }

    public List<Page> findPages(Chapter chapter) {
        Response pageResponse = japScanApiService.getReader(chapter.getMangaAlias(), chapter.getChapterNumber());

        String body = HttpUtils.convertFrom(pageResponse);

        List<Page> pages = parseMangaChapterPages(body);

        return pages;
    }

    public InputStream findPage(Page page) {
        Response pageResponse = japScanApiService.getPage(page.getMangaAlias(), page.getChapterNumber(), page.getPageNumber());

        String body = HttpUtils.convertFrom(pageResponse);
        Page finalPage = parseMangaChapterPage(body);

        page.setExtension(finalPage.getExtension());

        Response downloadPageResponse = japScanDataApiService.downloadPage(finalPage.getMangaAlias(), finalPage.getChapterNumber(), finalPage.getPageNumber(), finalPage.getExtension());

        return HttpUtils.readFrom(downloadPageResponse);
    }

    @Override
    public String getName() {
        return "JapScan";
    }

    @VisibleForTesting List<Manga> parseMangaList(String body) {
        List<Manga> mangas = new ArrayList<Manga>();

        Pattern pattern = Pattern.compile("<div class=\"row\"><div class=\"cell\"><a href=\"\\/mangas\\/([^\"]+)\\/\">([^<]+)<\\/a><\\/div><div class=\"cell\">([^<]+)<\\/div><div class=\"cell\">([^<]+)<\\/div><div class=\"cell\"><a href=\"([^\"]+)\">([^<]+)<\\/a><\\/div><\\/div>");
        Matcher matcher = pattern.matcher(body);
        while (matcher.find()) {
            String url = URLDecoder.decode(matcher.group(1));
            String name = matcher.group(2);

            Manga manga = createManga()
                    .withName(name)
                    .withMangaAlias(url)
                    .withProvider(getName())
                    .withLastChapter("")
                    .build();

            mangas.add(manga);
        }

        return mangas;
    }

    @VisibleForTesting List<Chapter> parseMangaChapters(String body) {
        List<Chapter> chapters = new ArrayList<Chapter>();

        Pattern pattern = Pattern.compile("<li><a href=\"\\/\\/www.japscan.com\\/lecture-en-ligne\\/([^\"]+)\\/([^\"]*)\\/\">([^<]*)<\\/a><\\/li>");
        Matcher matcher = pattern.matcher(body);
        while (matcher.find()) {
            String url = URLDecoder.decode(matcher.group(1));
            String chapterNumber = URLDecoder.decode(matcher.group(2));
            String name = matcher.group(3);

            Chapter chapter = createChapter()
                    .withName(name)
                    .withMangaAlias(url)
                    .withChapterNumber(chapterNumber)
                    .build();

            chapters.add(chapter);
        }

        return chapters;
    }

    @VisibleForTesting List<Page> parseMangaChapterPages(String body) {
        List<Page> pages = new ArrayList<Page>();

        Pattern pattern = Pattern.compile("<option value=\"\\/lecture-en-ligne\\/([^\"]+)\\/([^\"]+)\\/([^\"]+).html\"( selected=\"selected\")?>Page ([0-9]+)<\\/option>");
        Matcher matcher = pattern.matcher(body);
        while (matcher.find()) {
            String url = URLDecoder.decode(matcher.group(1));
            String chapterNumber = URLDecoder.decode(matcher.group(2));
            String pageNumber = URLDecoder.decode(matcher.group(3));

            Page page = createPage()
                    .withPageNumber(pageNumber)
                    .withMangaAlias(url)
                    .withChapterNumber(chapterNumber)
                    .build();

            pages.add(page);
        }

        return pages;
    }

    @VisibleForTesting Page parseMangaChapterPage(String body) {
        Pattern pattern = Pattern.compile("src=\"\\/\\/cdn.japscan.com\\/lecture-en-ligne\\/([^\"]+)\\/([^\"]+)\\/([^\".]+).([^\"]+)\"");
        Matcher matcher = pattern.matcher(body);
        while (matcher.find()) {
            String mangaAlias = URLDecoder.decode(matcher.group(1));
            String chapterNumber = URLDecoder.decode(matcher.group(2));
            String pageNumber = URLDecoder.decode(matcher.group(3));
            String extension = URLDecoder.decode(matcher.group(4).toLowerCase());

            Page page = createPage()
                    .withPageNumber(pageNumber)
                    .withMangaAlias(mangaAlias)
                    .withChapterNumber(chapterNumber)
                    .withExtenstion(extension)
                    .build();

            return page;
        }

        return null;
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/provider/japscan/downloader/JapScanDownloader.kt
package com.arnaudpiroelle.manga.provider.japscan.downloader


import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Rect
import com.arnaudpiroelle.manga.BuildConfig
import com.arnaudpiroelle.manga.core.provider.MangaProvider
import com.arnaudpiroelle.manga.core.rx.RxRequest
import com.arnaudpiroelle.manga.model.db.Manga
import com.arnaudpiroelle.manga.model.network.Chapter
import com.arnaudpiroelle.manga.model.network.Page
import com.arnaudpiroelle.manga.model.network.PostProcess
import io.reactivex.Single
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.Jsoup
import java.io.File
import java.io.FileOutputStream
import java.net.URLDecoder


class JapScanDownloader(private val okHttpClient: OkHttpClient) : MangaProvider {
    override fun findMangas(): Single<List<Manga>> {
        val request = Request.Builder()
                .url("${BuildConfig.JAPSCAN_BASE_URL}/mangas/")
                .build()

        return RxRequest(okHttpClient, request)
                .map(this::parseMangas)
                .onErrorReturn { listOf() }
    }

    override fun findChapters(manga: Manga): Single<List<Chapter>> {
        val request = Request.Builder()
                .url("${BuildConfig.JAPSCAN_BASE_URL}/mangas/${manga.mangaAlias}")
                .build()

        return RxRequest(okHttpClient, request)
                .map(this::parseChapters)
                .map { it.reversed() }
                .onErrorReturn { listOf() }

    }

    override fun findPages(manga: Manga, chapter: Chapter): Single<List<Page>> {
        val request = Request.Builder()
                .url("${BuildConfig.JAPSCAN_BASE_URL}/lecture-en-ligne/${manga.mangaAlias}/${chapter.chapterNumber}")
                .build()

        return RxRequest(okHttpClient, request)
                .map(this::parsePages)
                .onErrorReturn { listOf() }
    }

    override fun findPage(page: Page): Single<Response> {
        val request = Request.Builder()
                .url(page.url)
                .build()

        return RxRequest(okHttpClient, request)
    }

    private fun parseMangas(response: Response): List<Manga> {
        val doc = Jsoup.parse(response.body()?.string())
        val mangaElements = doc.select("div.cell>a[href*=/mangas/]")

        return mangaElements.map {
            val url = it.attr("href")
            val name = it.text()

            Manga(name, url.substringAfter("/mangas/"), "JapScan")
        }
    }

    private fun parseChapters(response: Response): List<Chapter> {
        val doc = Jsoup.parse(response.body()?.string())

        val chapters = doc.select("#liste_chapitres > ul > li:not(:has(span)) > a")
        return chapters.map {

            val url = it.attr("href")
            val name = it.text()

            val parts = url.split("/")
            val chapterNumber = URLDecoder.decode(parts[parts.lastIndex - 1], "UTF-8")
            val mangaAlias = URLDecoder.decode(parts[parts.lastIndex - 2], "UTF-8")

            Chapter(name, mangaAlias, chapterNumber)
        }

    }

    private fun parsePages(response: Response): List<Page> {
        val doc = Jsoup.parse(response.body()?.string())

        val playerJs = doc.select("script[src~=(lecteur.js|lecteur_cr.js).*]")

        val oldPlayer = playerJs.attr("src").contains("lecteur.js")
        val mangas = doc.getElementById("mangas")
        val chapitres = doc.getElementById("chapitres")
        val pages = doc.select("#pages > option:not([data-img~=(IMG__|__sy|__Add).*\\.(png|jpe?g)])")
        val image = doc.getElementById("image")

        val mangaName = mangas.dataset()["nom"]
        val chapterName = chapitres.dataset()["nom"]
        val chapterUri = chapitres.dataset()["uri"]

        return pages.filter { it.dataset()["img"]?.isNotEmpty() ?: false }
                .map { page ->
                    val img = page.dataset()["img"]
                    if (oldPlayer) {
                        val src = image.attr("src")
                        val index = src.lastIndexOf("/")
                        val substring = src.substring(0, index + 1)
                        Page(substring + img)
                    } else {
                        val cleanName = mangaName!!.replace("/", "_").replace("?", "")
                        val chapter = chapterName ?: chapterUri
                        val url = "${BuildConfig.JAPSCAN_CDN_BASE_URL}/cr_images/$cleanName/$chapter/$img"
                        Page(url, MosaicPostProcess)
                    }
                }

    }

    object MosaicPostProcess : PostProcess {
        override fun execute(file: File) {
            val source = BitmapFactory.decodeFile(file.absolutePath)
            val newBitmap = Bitmap.createBitmap(source.width, source.height, source.config)
            val comboImage = Canvas(newBitmap)

            val width = source.width.toDouble()
            val height = source.height.toDouble()

            val wp = Math.floor(width / 5)
            val hp = Math.floor(height / 5)
            val x = doubleArrayOf(wp * 2, wp * 4, 0.0, wp * 3, wp)
            val y = doubleArrayOf(hp * 4, hp * 3, hp * 2, hp, 0.0)

            var ph: Double?
            var pw: Double?

            for (i in 1..5) {
                ph = y[i - 1]

                for (j in 1..5) {
                    pw = x[j - 1]
                    val leftDst = (j - 1) * wp
                    val topDst = (i - 1) * hp
                    val rightDst = j * wp
                    val bottomDst = i * hp

                    val leftSrc = pw
                    val topSrc = ph
                    val rightSrc = pw + wp
                    val bottomSrc = ph + hp

                    val src = Rect(leftSrc.toInt(), topSrc.toInt(), rightSrc.toInt(), bottomSrc.toInt())
                    val dst = Rect(leftDst.toInt(), topDst.toInt(), rightDst.toInt(), bottomDst.toInt())

                    comboImage.drawBitmap(source, src, dst, null)
                }
            }

            FileOutputStream(file).use { os ->
                newBitmap.compress(Bitmap.CompressFormat.JPEG, 100, os)
                os.flush()
                os.close()
            }

            source.recycle()
        }
    }
}
